GeneratorInterface/RivetInterface/plugins/GenParticles2HepMCConverter.cc

```diff
@@ -196,7 +157,24 @@ void GenParticles2HepMCConverter::produce(edm::Event& event, const edm::EventSet
   }
 
   // Finalize HepMC event record
-  hepmc_event->set_signal_process_vertex(*(vertex1->vertices_begin()));
+  bool hasSignalVertex = false;
+  if ( !signalParticlePdgIds_.empty() ) {
+    // Loop over all vertices to assign the signal vertex, decaying to a signal particle
+    for ( auto v = hepmc_event->vertices_begin(); v != hepmc_event->vertices_end(); ++v ) {
+      for ( auto p = (*v)->particles_begin(HepMC::children);
+            p != (*v)->particles_end(HepMC::children); ++p ) {
+        const int pdgId = (*p)->pdg_id();
+        if ( signalParticlePdgIds_.count(pdgId) != 0 ) {
```

Comment:
@jhgoh 
here and elsewhere, count could be more efficiency replaced by 

signalParticlePdgIds_.find(pdgId) != signalParticlePdgIds_.end() 

if you just need to check that its present or not present (as opposed to counting how often a key is present)